### PR TITLE
[IMPROVE] Upgrade EmojiOne to JoyPixels 4.5.0

### LINKED ISSUE #13686: EmojiOne version 2 is no longer supported or distributed
### Description:
Rocket.Chat currently uses a very outdated version of EmojiOne. 
From https://github.com/emojione/emojione#emojione-version-2
>EmojiOne version 2 is no longer supported or distributed.

### Server Setup Information:
- Version of Rocket.Chat Server: 70afb0eff49131870e787b69bd0b009ac304653b

### PR DESCRIPTION
**Pull request by wreiske**
### This is READY TO MERGE!

For latest, see https://github.com/RocketChat/Rocket.Chat/pull/13807#issuecomment-480557910

Screen Shot 2019-03-20 at 3 40 17 AM

Screen Shot 2019-03-20 at 3 45 03 AM

Closes #13686

## Still TODO

- [x] Add "Emoji provided by JoyPixels" to the EmojiPicker as shown in https://github.com/RocketChat/Rocket.Chat/issues/13686#issuecomment-474606267
- [x] Fix reactions
- [x] i18n "Emoji Provided by JoyPixels"
- [x] ... anything else we notice that is broken.

# Notes
I modified the generateEmojiIndex script to automatically generate the emojiPicker.js file based off the emoji.json from emojione's github tagged 4.5.0.

Need to figure out how @sampaiodiego did the percentage mustache to allow resizing of the emojis without whitespace. **zoom** and **transform: scale(x)** do not work on all browsers (transform leaves whitespace)

**Firefox:**
image

**Webkit:** 
image

I'm thinking about including the 64x sprites and the 32x or 24x... The inline message emojis would use the 24x sprites, while the emoji replies (no other text) would use the 64x sprites (or maybe a size smaller.)

### CODE REVIEW

**Comment by wreiske** [comment]
Added attribution text in 4aa6564

image

**Comment by engelgabriel** [comment]
@wreiske this is awesome!

@sampaiodiego do you think we can use the assets directly from the npm module or do we have to copy them to `public/packages/emojione/assets/sprites/`? Ideally, we should avoid having to copy assets and keep them updated manually.

@rodrigok what do you think?

**Comment by rodrigok** [comment]
@engelgabriel AFAIK there is no way to provide assets from npm packages, we always need to copy them to the public folder.

**Comment by wreiske** [comment]
image

I fixed reactions in a local copy, but I'm not going to push it because it is counterproductive. Once we generate the percentages for the sprites, it will fix itself.

As for the assets, we could write a hook that copies the assets from the node modules to the public folder, or I could make an update_emojione.js that goes and pulls the latest sprites from their github. 🤔

**Comment by engelgabriel** [comment]
I guess we can use the SVG version?

**Comment by wreiske** [comment]
> I guess we can use the SVG version?

I think the free license they are modifying for us does not include high-quality SVG assets, just the PNG sprites.

**Comment by engelgabriel** [comment]
> I think the free license they are modifying for us does not include high-quality SVG assets, just the PNG sprites.

I agree, but maybe it is worth double checking?  😬

**Comment by wreiske** [comment]
> > I think the free license they are modifying for us does not include high-quality SVG assets, just the PNG sprites.
> 
> I agree, but maybe it is worth double checking? 😬

I'll send an email to Lisa and ask.

**Comment by engelgabriel** [comment]
Currently, we use a single sprite, right?
https://open.rocket.chat/packages/emojione_emojione/assets/sprites/emojione.sprites.png

I think that the two versions you have added could be enough, can you push your latest version so I can see on my mbp too?

**Comment by wreiske** [comment]
> Currently, we use a single sprite, right?
> https://open.rocket.chat/packages/emojione_emojione/assets/sprites/emojione.sprites.png
> 
> I think that the two versions you have added could be enough, can you push your latest version so I can see on my mbp too?

Pushed. This only has the 64 sprites in it. If you'd like, let me know the two sizes you would like and I'll get them committed. Once I push the other sprites, you'll have to make some code changes for them to display the different sizes. 

Change these in /app/emoji-emojione/lib/rocketchat.js
```
emoji.packages.emojione.emojiSize = 64;
emoji.packages.emojione.spriteSize = 64;
```

**Comment by wreiske** [comment]
Some things to consider about the SVG sprites...
https://demos.emojione.com/sprites-svg.html

>These are not supported in IE8.
>IE does not support SVG sprites, to get it working you'll need to include [SVG4Everybody](https://github.com/jonathantneal/svg4everybody)

**Comment by wreiske** [comment]
image

OK.. so this is with the 24x sprite sheet and the 64x sprite sheet without any size modifications. Had to make the reaction buttons a bit bigger.

big-reactions mov

**Comment by engelgabriel** [comment]
@ggazzo I think we can merge this and get to OPEN so people can test this.

**Comment by wreiske** [comment]
@engelgabriel pushed i18n. Please let me know if there is anything else you need from me. Thanks!

**Comment by rodrigok** [comment]
@wreiske you can automate the copy of the assets in `.scripts/npm-postinstall.js` file

**Comment by wreiske** [comment]
> .scripts/npm-postinstall.js

Sure thing. Give me a few...

**Comment by wreiske** [comment]
@rodrigok done. Not sure if we want to include the css too, since they require some modification. I could do a sed replace on the background image urls? for now, the assets are copying over.

**Comment by sampaiodiego** [comment]
> > > I think the free license they are modifying for us does not include high-quality SVG assets, just the PNG sprites.
> > 
> > 
> > I agree, but maybe it is worth double checking? grimacing
> 
> I'll send an email to Lisa and ask.

did you get a chance to ask them? :grimacing:

**Comment by wreiske** [comment]
> > > > I think the free license they are modifying for us does not include high-quality SVG assets, just the PNG sprites.
> > > 
> > > 
> > > I agree, but maybe it is worth double checking? grimacing
> > 
> > 
> > I'll send an email to Lisa and ask.
> 
> did you get a chance to ask them? 😬

Ahh, I forgot to share. 

image

That's why I moved forward with the two different sized PNGs.

**Comment by sampaiodiego** [comment]
thx @wreiske .. I was thinking if we should do like Discord. on the emoji picker they use a small (fixed size) png sprite, but on the messages they use an svg **per emoji**, they load the PNG using an `` tag.

so in this case we don't need a SVG sprite, but all individual emojis instead.

**Comment by sampaiodiego** [comment]
this is the ideal world (for me at least) and I was wondering if we could do the same.. if we can't do with SVG, maybe we could with PNG.. idk

**Comment by wreiske** [comment]
Do you want to merge this as is now? or should I attempt to make the reactions and inline message emojis use the individual PNGs? I doubt the SVG assets will be made available on the free plan...

**Comment by wreiske** [comment]
I'm hackin' this weekend, too! @engelgabriel let me know if there's any way I can help with this. 

👨‍💻

**Comment by engelgabriel** [comment]
Cool @wreiske! @sampaiodiego can you work with him to get this PR merged?

**Comment by sampaiodiego** [comment]
> Do you want to merge this as is now? or should I attempt to make the reactions and inline message emojis use the individual PNGs? I doubt the SVG assets will be made available on the free plan...

@wreiske I'm not sure we need to use individual PNGs for now.. but I have a other concerns, including a big performance decrease when handling emojis on the UI, either from emoji picker or emoji auto completion. Other are below:

image

* reactions are too big
* emojis are off center
* overall sizes are bigger:
  * before:
    * jumbomoji: 44x44
    * regular emoji: 22x22
    * reactions: 16x16
  * now:
    * jumbomoji: 64x64
    * regular emoji: 24x24
    * reactions: 24x24

image

* emojis are not showing on side nav's last message

**Comment by wreiske** [comment]
>reactions are too big
I agreee, the reactions are huge, but that was the smallest sprite sheet they offered. Without using individual PNGs, or changing the sprite sheet to use percentages (no idea how to do that), there is no good way to scale them down to the correct 16x16 reaction size from before. 
>emojis are off center
Since the in-line emoji is 2 pixels bigger, they may need some css to pull them down 2px.
overall sizes are bigger:

> before:
> jumbomoji: 44x44
> regular emoji: 22x22
> reactions: 16x16
> now:
> jumbomoji: 64x64
> regular emoji: 24x24
> reactions: 24x24
I enjoy the larger jumbomoji for single messages, but totally agree the regular emoji and reactions should be the same size they were before. I wonder if we could generate correct sized sprite sheets using something like https://draeton.github.io/stitches/ or https://www.npmjs.com/package/node-sprite-generator

I'd be happy to look into generating those.

>emojis are not showing on side nav's last message
Ahh, I knew I missed something! Those side navs will never work unless we have a much smaller sprite sheet generated or use the individual pngs.

**Comment by sampaiodiego** [comment]
thanks @wreiske .. the old `mustache` was used to generate old sprite's CSS (using percentages). I'll take a look on old emojione's repo and see if the script to generate the CSS is there (I think it is)

**Comment by wreiske** [comment]
> thanks @wreiske .. the old `mustache` was used to generate old sprite's CSS (using percentages). I'll take a look on old emojione's repo and see if the script to generate the CSS is there (I think it is)

I think I've got it figured out ;) Save your time. Should have something to show today.

**Comment by wreiske** [comment]
BOOM! @engelgabriel @sampaiodiego 

Screen Shot 2019-04-06 at 11 04 54 PM
Screen Shot 2019-04-06 at 11 04 26 PM

image

**Comment by wreiske** [comment]
**HOLD UP.** Apparently, I missed the tones!! **Don't merge this yet.** I'll get if fixed tomorrow.. unless someone wants to do it before then. Off to sleep now. I'll get this polished up tomorrow.

**Comment by wreiske** [comment]
Fixed the tones, too... now I think it's actually ready this time... 💤 😪 🛌 

Screen Shot 2019-04-07 at 2 21 18 AM
Screen Shot 2019-04-07 at 2 21 13 AM

**Comment by wreiske** [comment]
https://www.joypixels.com/blog/emojione-announces-company-name-change-to-joypixels
EmojiOne is now "JoyPixels"... Going to edit the README to fit.

**Comment by sampaiodiego** [comment]
awesome job @wreiske congratz... all emojis are indeed rendering as before =)

the only things that are still grabbing my attention are:

- overall emoji usage is waaay slower than before
- emoji order on emoji picker is different than before
- other very small thing (that would not prevent from merging) is the sprites path, I would change from `/packages/emojione/assets/sprites/flags-sprites.png` to something like `/emojione/flags-sprites.png`

the category order on the picker was also different, but I have fixed them now, I'll be sending a commit soon.

@engelgabriel what are your thoughts on first two items?

**Comment by wreiske** [comment]
One thing I noticed... When I added the attribution, I never fixed the height of the picker. 

Which of these do you like the most? I'll get it pushed.

### How it is currently in Rocket.Chat
image

### How it is now
image


### Update Height
**app/emoji/client/lib/EmojiPicker.js**
```
export const EmojiPicker = {
	width: 365,
	height: 305,
        ...,
```
Screen Shot 2019-04-08 at 11 44 40 AM

### Move Picker Above Input
**app/emoji/client/lib/EmojiPicker.js**
```
if (top + this.height >= windowHeight) {
    cssProperties.top = windowHeight - this.height - windowBorder - 75;
}
```
image

**Comment by geekgonecrazy** [comment]
Wow this is huge and those emoji look great!  Would be fantastic to get this as a bit of shine in 1.0 😄

**Comment by wreiske** [comment]
> Wow this is huge and those emoji look great! Would be fantastic to get this as a bit of shine in 1.0 😄

Digging in deep right now trying to fix the performance issues with the emoji picker.

**Comment by wreiske** [comment]
@sampaiodiego is this what you were thinking?

Apr-08-2019 18-03-59

**Comment by wreiske** [comment]
JoyPixels-Rocket-preview

Any thoughts on this new larger EmojiPicker + the category scrolling? @sampaiodiego I fixed the performance issues. 8467ms now takes 97ms!

**Comment by sampaiodiego** [comment]
@wreiske issues fixed. I couldn't reproduce the autocomplete and the re-render on each message though.

there is one issue still, when picking an emoji after searching throws an error regarding updating recent emojis.

**Comment by wreiske** [comment]
> @wreiske issues fixed. I couldn't reproduce the autocomplete and the re-render on each message though.
> 
> there is one issue still, when picking an emoji after searching throws an error regarding updating recent emojis.

I just pulled and tried your new changes. 👍 much better!

I also tried the autorun issue I was having before, and couldn't reproduce it.. so good there 👍 

One thing I just thought of, not a high priority.... if someone adds a new custom emoji, the emoji picker needs to re-render. Can you add something to the autorun that will trigger a re-render if a new emoji is added?

File: app/emoji-emojione/lib/generateEmojiIndex.mjs [hunk 1]
```diff
@@ -97,6 +95,7 @@ function generateEmojiPicker(data) {
 		output += `\t${category}: [\n`;
 
 		for (let emoji in emojisByCategory[category]) {
+			console.log('up', emojiList[emojisByCategory[category][emoji]]);
```
[the remark this answers is not in the document]
Comment (sampaiodiego) [reply app/emoji-emojione/lib/generateEmojiIndex.mjs]: oops

**Comment by sampaiodiego** [comment]
@wreiske about those two remaining issues, which are basically give some reactivity back to emoji picker, I'll be able to take look on them only tomorrow.
my idea was to change the code to instead of having a cache of everything, I would have a cache for each category (maybe a `ReactiveDict` instead of `ReactiveVar`), so if new custom or recent emoji comes, I would just update its category cache which would reactively update the UI.

**Comment by wreiske** [comment]
Sounds like a good plan. I'll keep my eyes out and pull/test anything I see new.

**Comment by wreiske** [comment]
@sampaiodiego anything I can do to help push this along? Let me know!

**Comment by sampaiodiego** [comment]
thanks @wreiske .. I was on vacations last week and rushing for 1.0 release previously.

I think this is almost ready for merge.. I'll take a look again this week.

**Comment by sampaiodiego** [comment]
alright.. this is now ready for merge!!

please @ggazzo and/or @tassoevan (or any other @RocketChat/core ) help reviewing.

thx a ton @wreiske

**Comment by wreiske** [comment]
Of course! Thank you for such an **AWESOME** open source project!

File: app/emoji/client/emojiPicker.js [hunk 2]
```diff
@@ -1,68 +1,51 @@
...
+	}).join('') || `<li>${ t('No_emojis_found') }</li>`;
+
 	return html;
+};
+
+export function updateRecentEmoji(category) {
+	emojiListByCategory.set(category, createEmojiList(category));
 }
 
+const createPickerEmojis = _.throttle((instance) => {
```
Comment (ggazzo) [inline app/emoji/client/emojiPicker.js]: check if you need this `throttle` here

File: app/emoji/client/emojiPicker.js [hunk 3]
```diff
@@ -182,17 +137,36 @@ Template.emojiPicker.events({
...
+		if (EmojiPicker.scrollingToCategory) {
+			return;
+		}
+
+		const container = instance.$(event.currentTarget);
+		const scrollTop = container.scrollTop() + 8;
+
+		const position = EmojiPicker.getCategoryPositions()
+			.filter((pos) => pos.top <= scrollTop)
+			.reverse();
```
Comment (ggazzo) [inline app/emoji/client/emojiPicker.js]: `array.pop`

File: app/emoji/client/emojiPicker.js [hunk 4]
```diff
@@ -230,7 +204,7 @@ Template.emojiPicker.events({
 
 		$('.tone-selector').toggleClass('show');
 	},
-	'click .emoji-list li'(event, instance) {
+	'click .emoji-list li.emoji-picker-item'(event, instance) {
```
Comment (ggazzo) [inline app/emoji/client/emojiPicker.js]: `click .emoji-list .emoji-picker-item`

File: app/emoji/client/emojiPicker.js [hunk 5]
```diff
@@ -281,8 +267,10 @@ Template.emojiPicker.onCreated(function() {
 	};
 
 	this.autorun(() => {
-		if (this.recentNeedsUpdate.get()) {
-			this.recentNeedsUpdate.set(false);
+		if (recentEmojisNeedsUpdate.get()) {
```
Comment (ggazzo) [inline app/emoji/client/emojiPicker.js]: 🤔 could you help me to understand what this code does?

File: app/emoji/client/lib/EmojiPicker.js [hunk 6]
```diff
@@ -114,33 +128,57 @@ export const EmojiPicker = {
...
+		Tracker.afterFlush(() => {
+			const header = $(`#emoji-list-category-${ category }`);
+			const container = $('.emoji-picker .emojis');
+
+			const scrollTop = header.position().top + container.scrollTop();// - container.position().top;
+
+			if (animate) {
+				return container.animate({
+					scrollTop,
+				}, 300, () => setTimeout(() => { this.scrollingToCategory = false; }, 200));
```
Comment (ggazzo) [inline app/emoji/client/lib/EmojiPicker.js]: we dont need `setTimeout`

File: app/emoji/client/emojiPicker.js (hunk 5 above)
Comment (sampaiodiego) [reply app/emoji/client/emojiPicker.js]: legacy cold

**Comment by wreiske** [comment]
:tada: :rocket:

**Comment by alyssais** [comment]
Reposting [my comment](https://github.com/RocketChat/Rocket.Chat/issues/13686#issuecomment-485876309) from #13686 here in case somebody in this thread can answer:

> Is there a copy of the full license that will apply to users of rocketchat after this upgrade? What terms does it impose on redistributors?

If I host a Rocket.Chat instance, or create a package for it, under what conditions am I permitted to use the JoyPixels emoji? Presumably I can't redistribute them freely or modify them, but without any sort of license I can't know whether I'm allowed to use them at all!

**Comment by wreiske** [comment]
> Reposting [my comment](https://github.com/RocketChat/Rocket.Chat/issues/13686#issuecomment-485876309) from #13686 here in case somebody in this thread can answer:
> 
> > Is there a copy of the full license that will apply to users of rocketchat after this upgrade? What terms does it impose on redistributors?
> 
> If I host a Rocket.Chat instance, or create a package for it, under what conditions am I permitted to use the JoyPixels emoji? Presumably I can't redistribute them freely or modify them, but without any sort of license I can't know whether I'm allowed to use them at all!

I have reached out to the Joypixels team for comment. I'll let you know their response.

**Comment by wreiske** [comment]
@alyssais 
Just an update: I have connected Lisa from JoyPixels with Gabriel. They are still working on drafting the agreement for Rocket.Chat's use of JoyPixels. Their licensing department is a bit backed-up, but they should have an amended user agreement to look at next week sometime.

**Comment by alyssais** [comment]
> @alyssais Just an update: I have connected Lisa from JoyPixels with
> Gabriel. They are still working on drafting the agreement for
> Rocket.Chat's use of JoyPixels. Their licensing department is a bit
> backed-up, but they should have an amended user agreement to look at
> next week sometime.

Guessing you never heard anything?

**Comment by wreiske** [comment]
> @alyssais Just an update: I have connected Lisa from JoyPixels with Gabriel. They are still working on drafting the agreement for Rocket.Chat's use of JoyPixels. Their licensing department is a bit backed-up, but they should have an amended user agreement to look at next week sometime.
> Guessing you never heard anything?

Actually, they responded to @engelgabriel I believe. Maybe we need to include the license here @engelgabriel ?

**Comment by heyyo-droid** [comment]
Reducing the palette to 256 colors, could reduce the sprite size to 1.05MB vs 3.455MB.
people-sprites

**Pull request merged**